lib/broadway/producer.ex

```diff
@@ -236,4 +266,75 @@ defmodule Broadway.Producer do
             "struct. In case you're using a standard GenStage producer, please set the " <>
             ":transformer option to transform produced events into message structs"
   end
+
+  defp maybe_rate_limit_and_buffer_messages(state, messages) do
+    if state.rate_limiting do
+      state = update_in(state.rate_limiting.message_buffer, &:queue.in(messages, &1))
```

Comment:
Not a big of this this approach because of the overhead. We are putting it in the queue only to take it out in most cases. Maybe we should have something like `rate_limit_and_buffer_messages(messages, state)`.